File: heightBalanced.cpp
// Time Complexity :O(n^2) where n is the number of nodes(worst case)
// Space Complexity : O(h) Height of the tree  
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        if(!root) return true;
        // findLeftHeight
        int leftHeight = findHeight(root->left);
        //findRightHeight
        int rightHeight = findHeight(root->right);
        //Compare and return if false ; if not continue
        if(abs(rightHeight - leftHeight) > 1) return false;
        return isBalanced(root->left) && isBalanced(root->right);    
    }
    // Height func()
    int findHeight(TreeNode* root){
        if(!root) return 0;
        int left = findHeight(root->left);
        int right = findHeight(root-> right);
        return max(left, right) + 1;
    }
};

File: HeightBalanced.java
//Brute force-- expensive method
//public class HeightBalanced {
//    public boolean isBalanced(TreeNode root) {
//        if (root==null) return true;
//        if(Math.abs(height(root.left)-height(root.right))>1)return false;
//        return isBalanced(root.left) && isBalanced(root.right);
//
//    }
//    private int height(TreeNode root){
//        if(root==null) return 0;
//        return Math.max(height(root.left),height(root.right)) + 1;
//    }
//}

//Optimized
public class HeightBalanced {
    public boolean isBalanced(TreeNode root) {
        if (root==null) return true;
        return height(root)!=-1;

    }

    private int height(TreeNode root){
        if(root==null) return 0;
        int left = height(root.left);
        int right = height(root.right);
        if(Math.abs(left-right)>1) return -1;
        if(left==-1||right==-1) return -1;
        return Math.max(left,right) + 1;
    }
}
